CardMatchingGameWithBean/Model/CardMatchingGame.m

```diff
@@ -45,48 +49,93 @@ -(instancetype) initWithCardCount:(NSUInteger)count usingDeck:(Deck *)deck
+    // to the specific match game.
+
+    Card *card = [self cardAtIndex:cardIndex];
+    [self.chosenCards addObject:card];
+    [self scoreGame:@"picked a card"];
     
-    if (!card.isMatched) {
-        if (card.isChosen) {
+    if (!card.matched) {
+        if (card.chosen) {
             card.chosen = NO;
+            [self.chosenCards removeObject:card];
         } else {
-            for (Card *otherCard in self.cards) {
-                if (otherCard.isChosen && !otherCard.isMatched) {
-                    int matchScore = [card match:@[otherCard]];
-                    if (matchScore) {
-                        self.score += (matchScore * MATCH_BONUS);
-                        card.matched = YES;
-                        otherCard.matched = YES;
-                    } else {
-                        self.score -= MISMATCH_PENALTY;
-                        otherCard.chosen = NO;
-                    }
-                    break;
+            NSInteger cardsInArray =
+                [self _differentiateTwoAndThreeCardGames:
+                          segmentedControlIndex];
+            if (cardsInArray) {
+                self.matchScore = [card match:self.chosenCards];
+                if (self.matchScore) {
+                    [self matchCards:self.chosenCards];
+                } else if (self.matchScore == 0) {
+                    [self unmatchCards:self.chosenCards];
                 }
             }
-            
-            self.score -= COST_TO_CHOOSE;
             card.chosen = YES;
         }
     }
 }
 
+static const int MISMATCH_PENALTY = 2;
+static const int MATCH_BONUS = 4;
+static const int COST_TO_CHOOSE = 1;
+
+-(void)scoreGame:(NSString *)indicator;
+{
+    if ([indicator isEqualToString:@"picked a card"]) {
+        self.score -= COST_TO_CHOOSE;
+    } else if ([indicator isEqualToString:@"cards matched"]) {
+        self.score += (self.matchScore * MATCH_BONUS);
+    } else if ([indicator isEqualToString:@"cards don't match"]) {
+        self.score -= MISMATCH_PENALTY;
+    }
+}
+
+-(void)matchCards: (NSMutableArray* ) userCards
+{
+    for (int i = 0; (i <= [userCards count]- 1); i++) {
+        Card* cardToBeMatched = userCards[i];
+        [self scoreGame: @"cards matched"];
+        cardToBeMatched.matched = YES;
+    }
+    self.chosenCards = [NSMutableArray array];
+}
 
+-(void) unmatchCards: (NSMutableArray *) usersCards
+{
+    for (int i = 0; (i < [usersCards count]-1); i++) {
+        Card* cardToBeUnMatched = usersCards[i];
+        [self scoreGame:@"cards don't match"];
+        [self.chosenCards removeObject:cardToBeUnMatched];
+        NSLog(@"CARDS AFTER BEING REMOVED %ld", [self.chosenCards count]);
+        cardToBeUnMatched.chosen = NO;
+    }
+}
 
+- (NSInteger)_differentiateTwoAndThreeCardGames:(NSInteger)segmentControlIndex
+{
+    return [self.chosenCards count] == segmentControlIndex + 2 ? [self.chosenCards count]: 0;
```

Comment:
In order for me to set the segmentControlIndex properly, I'll have to add the cards in the array in the ViewController, and put the matching logic there (how cards are being flipped as they pick it.).  As the user are picking cards, the cards are flipping over.  If I add the cards in an array in the ViewController, then handle the flipping logic, the cards won't flip when the pick it. It'll flip afterwards.